Upgrade node-mssql to ^4.2.1

gshively11 opened the pull request:
- node-mssql 3.x->4.x has breaking interface changes, see 0.3.0 in our CHANGELOG.md for more information
- Improved DEBUG logs
- Fixed/improved tests
- Added some patches for existing bugs in node-mssql 4.2.1
- ~Idle pool connections close within 500ms by default, instead of the previous default of 30000ms~ edit: reverted this change
- Stats have changed slightly, due to changes in node-mssql internals
- Added package-lock.json
- Minor doc changes

This upgrade is preparation for Availability Group support. node-mssql 4.x uses a newer version of the tedious driver which should support AG's. This change alone may be sufficient for AG support, but further tests are still needed to confirm this.

(The addition of the package-lock.json accounts for the majority of line changes, don't let the big number deter you from reviewing the PR 😄)

gshively11 committed:
Upgrade node-mssql to ^4.2.1

- node-mssql 3.x->4.x has breaking interface changes, see 0.3.0 in our CHANGELOG.md for more information
- Improved DEBUG logs
- Fixed/improved tests
- Added some patches for existing bugs in node-mssql 4.2.1
- Idle pool connections close within 500ms by default, instead of the previous default of 30000ms
- Stats have changed slightly, due to changes in node-mssql interals
- Added package-lock.json
- Minor doc changes

bmeck commented on README.md:
```diff
@@ -105,6 +118,8 @@ export default function runQuery(id) {
 
 Check out the [detailed API documentation](API.md#new-connectionpoolpartyconfig).
 
+You'll also want to familiarize yourself with [`node-mssql`'s documentation](https://github.com/tediousjs/node-mssql/blob/7f374a8d73b00b17aa5b5ea5621c4314fc6e2daa/README.md#documentation), as much of it applies to mssql-pool-party.
```
should we be linking to master here, or are we going to keep it pinned?

gshively11 replied:
hmm, i suppose we should point it at master

pnevares commented on src/request-method-failure.js:
```diff
@@ -4,7 +4,7 @@ const debug = setDebug('mssql-pool-party');
 export default function requestMethodFailure(request, attempts, cb) {
   return (err) => {
-    debug('request failed!');
+    debug(`request ${request.id}failed!`);
```
nitpick: spacing

bmeck commented on src/connection-pool-party.js:
```diff
@@ -461,7 +484,9 @@ export default class ConnectionPoolParty extends EventEmitter {
     // get any updated dsn info from the provider
-    this._healingPromise = this._healingPromise || this.dsnProvider()
+    this._healingPromise = this._healingPromise || Promise.resolve()
+      .then(() => debug('healing started, retrieving new dsns from provider'))
```
wouldn't these only apply to the `Promise.resolve()`? the `||` needs some parens it looks like.

gshively11 replied:
that's the intent here. it's a mechanism to ensure the healing code is only run once but still allow multiple calls to wait for the resolution

bmeck replied:
just confusing to read, putting () around the one it is intended to apply to would at least show it is intentional

gshively11 replied:
ah i see, yea that's reasonable, will add

bmeck reviewed (approved):
only non-blocking nits

asilvas commented on API.md:
```diff
@@ -41,7 +41,9 @@ reconnets/retries, and basic health/statistics reporting.
 | [config.dsnProvider] | <code>function</code> |  | A function returning a promise that resolves  with an array of dsn object(s). This option will override config.dsn and config.dsns.  Required if dsn and dsns are not provided. |
 | [config.connectionPoolFactory] | <code>function</code> |  | A function that receives the dsn objects  from the dnsProvider and returns a promise that resolves with *connected* instance(s) of  ConnectionPool. Use this option if you want to customize how mssql ConnectionPools are  instantiated and connected. |
-| [config.connectionPoolConfig] | <code>object</code> |  | An object containing any configuration  you want to attach to the config provided when creating an mssql ConnectionPool. This is  useful if you don't want to create a custom dsnProvider or connectionPoolFactory to modify  the configuration used to create ConnectionPools. Just keep in mind that any config set here  will override the config ...
+| [config.connectionPoolConfig] | <code>object</code> |  | An object containing any configuration  you want to attach to the config provided when creating an mssql ConnectionPool. This is  useful if you don't want to create a custom dsnProvider or connectionPoolFactory to modify  the configuration used to create ConnectionPools. Just keep in mind that any config set here  will override the config ...
```
Link to docs would be handy

asilvas commented on CHANGELOG.md:
````diff
@@ -1,3 +1,62 @@
+  rowsAffected: [],
+}
+```
+- Pool inside `node-mssql`'s ConnectionPool now defaults both `evictionRunIntervalMillis` and `idleTimeoutMillis` to `500ms` (original defaults were `1000ms` and `30000ms` respectively). Can be overridden by modifying `connectionPoolConfig.pool.X`.
````
Seems overly aggressive and wasteful defaults? What was the motivation for such low defaults?

gshively11 replied:
i made that change [because of this](https://github.com/tediousjs/node-mssql/issues/457).  basically, node-mssql wouldn't actually close it's pool connections, which caused jest to hang.  but you have a point, i shouldn't make that the default for everyone, only for the tests.  will fix, thanks for catching that

asilvas commented on package.json:
```diff
@@ -9,13 +9,15 @@
     "docs": "jsdoc2md --configure jsdoc.json src/**/*.js > API.md",
     "lint": "node_modules/.bin/eslint src test --fix",
-    "test": "npm run lint && node_modules/.bin/jest",
+    "test": "npm run lint && node_modules/.bin/jest src/*",
```
Scripts use `node_modules/.bin` for pathing, so you should be able to `npm run lint && jest src/*`. I assume it works with multiple commands as well.

jcrugzz commented on CHANGELOG.md:
```diff
@@ -1,3 +1,62 @@
+# 0.3.0 (TBD)
```
Technically this module doesn't have breaking changes but the underlying `node-mssql` does. It might be worth finally going to `1.0.0` to properly indicate the breaking change so major can be used to indicate breaking changes in the future. Thoughts @gshively11?

gshively11 replied:
for `0.x.x` versions, the minor is technically a "major" bump right?  e.g. people with `^0.2.0` won't automatically get `0.3.0`.  i'm not against going to `1.0.0`, but i'd rather do that when i've confirmed AGs work as intended

jcrugzz replied:
Pre `1.0.0` you are safe from `^` bumping a major version in semver since `^ === ~` until you hit `1.0.0`. People tend to limit the time in pre `1.0.0` land due to the inconsistent semantics though. I think its fair to confirm AGs work as you expect and then pull the trigger. Just wanted to encourage getting to `1.0.0` since it is production ready software :).

gshively11 replied:
^ you make a good point, will bump for the next release

gshively11 committed:
Addressed PR feedback

- Improved code clarity
- Improved documentation
- Reverted the pool timeout settings in the previous commit, now follows the same defaults as `node-mssql`
- Updated pool timeouts for tests to workaround bug
- Removed unnecessary pathing in package.json scripts
